File: app/playRepository/Commit.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2012 NAVER Corp.
 * http://yobi.io
 *
 * @author Yi EungJun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package playRepository;

import models.*;
import models.enumeration.ResourceType;
import models.resource.Resource;

import java.util.*;

public abstract class Commit {

    public abstract String getShortId();
    public abstract String getId();
    public abstract String getShortMessage();
    public abstract String getMessage();
    public abstract User getAuthor();
    public abstract String getAuthorName();
    public abstract String getAuthorEmail();
    public abstract Date getAuthorDate();
    public abstract TimeZone getAuthorTimezone();
    public abstract String getCommitterName();
    public abstract String getCommitterEmail();
    public abstract Date getCommitterDate();
    public abstract TimeZone getCommitterTimezone();
    public abstract int getParentCount();

    public Set<User> getWatchers(Project project) {
        Set<User> actualWatchers = new HashSet<>();

        // Add the author
        if (!getAuthor().isAnonymous()) {
            actualWatchers.add(getAuthor());
        }

        // Add every user who comments on this commit
        if (project.vcs.equals(RepositoryService.VCS_GIT)) {
            List<CommentThread> threads = CommentThread.find.where()
                    .eq("project.id", project.id)
                    .eq("commitId", getId())
                    .eq("pullRequest.id", null).findList();
            for (CommentThread thread : threads) {
                for (ReviewComment comment : thread.reviewComments) {
                    User user = User.find.byId(comment.author.id);
                    if (user != null) {
                        actualWatchers.add(user);
                    }
                }
            }
        } else {
            List<CommitComment> comments = CommitComment.find.where()
                    .eq("project.id", project.id).eq("commitId", getId()).findList();
            for (CommitComment c : comments) {
                User user = User.find.byId(c.authorId);
                if (user != null) {
                    actualWatchers.add(user);
                }
            }
        }

        return Watch.findActualWatchers(actualWatchers, asResource(project));
    }

    public static Project getProjectFromResourceId(String resourceId) {
        String[] pair = resourceId.split(":");
        return Project.find.byId(Long.valueOf(pair[0]));
    }

    public static Resource getAsResource(Project project, String commitId) {
        return getAsResource(project.id + ":" + commitId);
    }

    public static Resource getAsResource(final String resourceId) {
        return new Resource() {

            @Override
            public String getId() {
                return resourceId;
            }

            @Override
            public Project getProject() {
                return getProjectFromResourceId(resourceId);
            }

            @Override
            public ResourceType getType() {
                return ResourceType.COMMIT;
            }
        };
    }

    public Resource asResource(final Project project) {
        return new Resource() {

            @Override
            public String getId() {
                return project.id + ":" + Commit.this.getId();
            }

            @Override
            public Project getProject() {
                return project;
            }

            @Override
            public ResourceType getType() {
                return ResourceType.COMMIT;
            }

            @Override
            public Long getAuthorId() {
                User author = getAuthor();

                if (author != null) {
                    return getAuthor().id;
                } else {
                    return null;
                }
            }
        };
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/vcs/Commit.kt
package com.github.yonaprojects.yona.domain.vcs

import com.github.yonaprojects.yona.domain.gpgkey.GpgVerificationStatus
import com.github.yonaprojects.yona.domain.user.User
import java.util.Date
import java.util.TimeZone

abstract class Commit {
    abstract fun getShortId(): String
    abstract fun getId(): String
    abstract fun getShortMessage(): String
    abstract fun getMessage(): String?
    abstract fun getAuthor(): User?
    abstract fun getAuthorName(): String?
    abstract fun getAuthorEmail(): String?
    abstract fun getAuthorDate(): Date?
    abstract fun getAuthorTimezone(): TimeZone?
    abstract fun getCommitterName(): String?
    abstract fun getCommitterEmail(): String?
    abstract fun getCommitterDate(): Date?
    abstract fun getCommitterTimezone(): TimeZone?
    abstract fun getParentCount(): Int

    // 커밋은 이 앱에서 JPA 엔티티로 미러링되지 않고 항상 git/svn 저장소
    // 원본에서 그때그때 읽어오므로, GPG 검증
    // 결과도 DB 캐시 컬럼이 아니라 조회 시점에 즉시 계산해 반환한다(GitCommit 참고). SVN은 GPG
    // 서명 개념이 없으므로 SvnCommit은 항상 UNSIGNED를 반환한다.
    abstract fun getGpgVerificationStatus(): GpgVerificationStatus
}
